File: src/main/java/umg/edu/progra/bst/BST.ts
package umg.edu.progra.bst;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 *
 * @author Walter Cordova
 */
class BST<T extends Empleado> {
    private valor: T | null;
    private izdo: BST<T> | null;
    private dcho: BST<T> | null;
    private padre: BST<T> | null;

    constructor() {
        this.valor = null;
        this.izdo = null;
        this.dcho = null;
        this.padre = null;
    }

    insertar(emp: T, padre: BST<T> | null = null): void {
        if (this.valor === null) {
            this.valor = emp;
            this.padre = padre;
        } else {
            if (emp.compareTo(this.valor) < 0) {
                if (this.izdo === null) {
                    this.izdo = new BST<T>();
                }
                this.izdo.insertar(emp, this);
            } else if (emp.compareTo(this.valor) > 0) {
                if (this.dcho === null) {
                    this.dcho = new BST<T>();
                }
                this.dcho.insertar(emp, this);
            } else {
                throw new Error("Insertando elemento duplicado");
            }
        }
    }

    existe(id: number): boolean {
        if (this.valor !== null) {
            if (id === this.valor.getId()) {
                return true;
            } else if (this.izdo !== null && id < this.valor.getId()) {
                return this.izdo.existe(id);
            } else if (this.dcho !== null && id > this.valor.getId()) {
                return this.dcho.existe(id);
            } else {
                return false;
            }
        } else {
            return false;
        }
    }

    obtener(id: number): T | null {
        if (this.valor !== null) {
            if (id === this.valor.getId()) {
                return this.valor;
            } else if (this.izdo !== null && id < this.valor.getId()) {
                return this.izdo.obtener(id);
            } else if (this.dcho !== null && id > this.valor.getId()) {
                return this.dcho.obtener(id);
            } else {
                return null;
            }
        } else {
            return null;
        }
    }

    esHoja(): boolean {
        return this.valor !== null && this.izdo === null && this.dcho === null;
    }

    esVacio(): boolean {
        return this.valor === null;
    }

    preOrden(): void {
        this.preordenImpl("");
    }

    postOrden(): void {
        this.postordenImpl("");
    }

    inOrden(): void {
        this.inordenImpl("");
    }

    private inordenImpl(prefijo: string): void {
        if (this.izdo !== null) {
            this.izdo.inOrden();
        }
        console.log(prefijo + this.valor);
        if (this.dcho !== null) {
            this.dcho.inOrden();
        }
    }

    private postordenImpl(prefijo: string): void {
        if (this.izdo !== null) {
            this.izdo.postOrden();
        }
        if (this.dcho !== null) {
            this.dcho.postOrden();
        }
        console.log(prefijo + this.valor);
    }

    private preordenImpl(prefijo: string): void {
        console.log(prefijo + this.valor);
        if (this.izdo !== null) {
            this.izdo.preOrden();
        }
        if (this.dcho !== null) {
            this.dcho.preOrden();
        }
    }

    eliminar(id: number): void {
        if (this.valor !== null) {
            if (id === this.valor.getId()) {
                this.eliminarImpl(id);
            } else if (this.izdo !== null && id < this.valor.getId()) {
                this.izdo.eliminar(id);
            } else if (this.dcho !== null && id > this.valor.getId()) {
                this.dcho.eliminar(id);
            }
        }
    }

    private eliminarImpl(id: number): void {
        if (this.izdo !== null && this.dcho !== null) {
            let sucesor: BST<T> | null = this.dcho;
            while (sucesor.izdo !== null) {
                sucesor = sucesor.izdo;
            }
            this.valor = sucesor.valor;
            this.dcho.eliminar(this.valor.getId());
        } else if (this.izdo !== null || this.dcho !== null) {
            const hijo: BST<T> = (this.izdo !== null) ? this.izdo : this.dcho!;
            if (this.padre !== null) {
                if (this === this.padre.izdo) {
                    this.padre.izdo = hijo;
                } else {
                    this.padre.dcho = hijo;
                }
                hijo.padre = this.padre;
            } else {
                this.valor = hijo.valor;
                this.izdo = hijo.izdo;
                this.dcho = hijo.dcho;
            }
        } else {
            if (this.padre !== null) {
                if (this === this.padre.izdo) {
                    this.padre.izdo = null;
                } else {
                    this.padre.dcho = null;
                }
            } else {
                this.valor = null;
            }
        }
    }
}

File: src/main/java/umg/edu/progra/bst/BST.java
package umg.edu.progra.bst;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 *
 * @author Walter Cordova
 */
public class BST implements IBST<Empleado> {

    private static final Logger LOGGER = LoggerFactory.getLogger(BST.class);

    private Empleado valor;
    private BST izdo, dcho;
    private BST padre;

//    @Override
//    public void insertar(Empleado emp) {
//        if (valor == null) {
//            this.valor = emp;
//        } else {
//            if (emp.compareTo(valor) < 0) {
//                if (izdo == null) {
//                    izdo = new BST();
//                }
//                izdo.insertar(emp);
//            } else if (emp.compareTo(valor) > 0) {
//                if (dcho == null) {
//                    dcho = new BST();
//                }
//                dcho.insertar(emp);
//            } else {
//                throw new RuntimeException("Insertando elemento duplicado");
//            }
//        }
//    }
    private void insertarImp(Empleado emp, BST padre) {
        if (valor == null) {
            this.valor = emp;
            this.padre = padre;
        } else {
            if (emp.compareTo(valor) < 0) {
                if (izdo == null) {
                    izdo = new BST();
                }
                izdo.insertarImp(emp, this);
            } else if (emp.compareTo(valor) > 0) {
                if (dcho == null) {
                    dcho = new BST();
                }
                dcho.insertarImp(emp, this);
            } else {
                throw new RuntimeException("Insertando elemento duplicado");
            }
        }
    }

    @Override
    public void insertar(Empleado emp) {
        insertarImp(emp, null);
    }

    @Override
    public boolean existe(int id) {
        if (valor != null) {
            if (id == valor.getId()) {
                return true;
            } else if (izdo != null && id < valor.getId()) {
                return izdo.existe(id);
            } else if (dcho != null && id > valor.getId()) {
                return dcho.existe(id);
            } else {
                return false;
            }
        } else {
            return false;
        }
    }

    @Override
    public Empleado obtener(int id) {
        if (valor != null) {
            if (id == valor.getId()) {
                return valor;
            } else if (izdo != null && id < valor.getId()) {
                return izdo.obtener(id);
            } else if (dcho != null && id > valor.getId()) {
                return dcho.obtener(id);
            } else {
                return null;
            }
        } else {
            return null;
        }
    }

    @Override
    public boolean esHoja() {
        return valor != null && izdo == null && dcho == null;
    }

    @Override
    public boolean esVacio() {
        return valor == null;
    }

    @Override
    public void preOrden() {
        preordenImpl("");
    }

    @Override
    public void postOrden() {
        postordenImpl("");
    }

    @Override
    public void inOrden() {
        inordenImpl("");
    }

    private void inordenImpl(String prefijo) {
        if (izdo != null) {
            izdo.inordenImpl(prefijo + "  ");
        }
        LOGGER.info("{}", prefijo + valor);
        if (dcho != null) {
            dcho.inordenImpl(prefijo + "  ");
        }
    }

    private void postordenImpl(String prefijo) {
        if (izdo != null) {
            izdo.postordenImpl(prefijo + "  ");
        }
        if (dcho != null) {
            dcho.postordenImpl(prefijo + "  ");
        }
        LOGGER.info("{}", prefijo + valor);
    }

    private void preordenImpl(String prefijo) {
        if (valor != null) {
            LOGGER.info("{}", prefijo + valor);
            if (izdo != null) {
                izdo.preordenImpl(prefijo + "  ");
            }
            if (dcho != null) {
                dcho.preordenImpl(prefijo + "  ");
            }
        }
    }

    private void eliminarImpl(int id) {
        if (izdo != null && dcho != null) {
            // Caso 1: el nodo tiene dos hijos
            // Encontrar el sucesor inmediato en el subárbol derecho
            BST sucesor = dcho;
            while (sucesor.izdo != null) {
                sucesor = sucesor.izdo;
            }
            // Copiar el valor del sucesor al nodo actual
            valor = sucesor.valor;
            // Eliminar el sucesor recursivamente
            dcho.eliminarImpl(sucesor.valor.getId());
        } else if (izdo != null || dcho != null) {
            // Caso 2: el nodo tiene un solo hijo
            BST hijo = (izdo != null) ? izdo : dcho;
            if (padre != null) {
                // Enlazar el hijo con el padre del nodo actual
                if (this == padre.izdo) {
                    padre.izdo = hijo;
                } else {
                    padre.dcho = hijo;
                }
                hijo.padre = padre;
            } else {
                // El nodo actual es la raíz
                valor = hijo.valor;
                izdo = hijo.izdo;
                dcho = hijo.dcho;
            }
        } else {
            // Caso 3: el nodo es una hoja
            if (padre != null) {
                // Eliminar el nodo según su posición respecto al padre
                if (this == padre.izdo) {
                    padre.izdo = null;
                } else {
                    padre.dcho = null;
                }
            } else {
                // El nodo es la raíz del árbol
                valor = null;
            }
        }
    }

    @Override
    public void eliminar(int id) {
        if (valor != null) {
            if (id == valor.getId()) {
                //eliminar valor
                eliminarImpl(id);
            } else if (izdo != null && id < valor.getId()) {
                izdo.eliminarImpl(id);
            } else if (dcho != null && id > valor.getId()) {
                dcho.eliminarImpl(id);
            }
        }
    }

}
